server/aesdsocket.c

```diff
@@ -15,23 +13,186 @@
 #include <fcntl.h>
 #include <signal.h>
 #include <errno.h>
+#include <pthread.h>
+#include <sys/queue.h>
+#include <stdbool.h>
 #define FAILURE -1
 #define BACKLOG 10 // how many pending connections queue will hold
 #define PORT_NUMBER "9000"
-#define FILE_LOG_PATH "/var/tmp/aesdsocketdata" //path to output file
+#define FILE_LOG_PATH "/var/tmp/aesdsocketdata" // path to output file
 
 int socket_fd;
-//signal handler
+bool sig_recv = false;
+pthread_mutex_t mutex_socket;
+pthread_t timestamp_thread;
+// signal handler
 void signal_handler(int sig_num)
 {
     syslog(LOG_USER, "Caught Signal : %s\n", strsignal(sig_num));
     close(socket_fd);
     remove(FILE_LOG_PATH);
+    sig_recv = true; // indicate signal caught
     closelog();
     exit(0);
```

Comment:
Not safe from signal handler, you should use `sig_rev` to shut down cleanly instead.